OpenRA.Mods.Common/UpdateRules/Rules/20221203/RemoveNegativeSequenceLength.cs

```diff
@@ -0,0 +1,86 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class RemoveNegativeSequenceLength : UpdateRule
+	{
+		public override string Name => "Negative sequence length is no longer allowed.";
+
+		public override string Description => "Negative sequence length is no longer allowed, define individual frames in reverse instead.";
+
+		List<MiniYamlNode> resolvedImagesNodes;
+
+		public override IEnumerable<string> BeforeUpdateSequences(ModData modData, List<MiniYamlNode> resolvedImagesNodes)
+		{
+			this.resolvedImagesNodes = resolvedImagesNodes;
+			yield break;
+		}
+
+		readonly Queue<Action> actionQueue = new();
+
+		public override IEnumerable<string> UpdateSequenceNode(ModData modData, MiniYamlNode imageNode)
+		{
+			foreach (var node in imageNode.Value.Nodes)
+			{
+				var lengthNode = node.LastChildMatching("Length", includeRemovals: false);
+				if (lengthNode == null || lengthNode.Value.Value == "*")
+					continue;
+
+				var length = FieldLoader.GetValue<int>(lengthNode.Key, lengthNode.Value.Value);
+				if (length >= 0)
+					continue;
+
+				var startNode = node.LastChildMatching("Start", includeRemovals: false);
+				if (startNode == null && node.Value.Value != "Defaults")
+					startNode = imageNode.LastChildMatching("Defaults")?.LastChildMatching("Start", includeRemovals: false);
+
+				if (startNode == null)
+				{
+					var resolvedImage = resolvedImagesNodes.First(n => n.Key == imageNode.Key);
+					startNode = resolvedImage.LastChildMatching(node.Key)?.LastChildMatching("Start", includeRemovals: false);
+
+					if (startNode == null && node.Value.Value != "Defaults")
+						startNode = resolvedImage.LastChildMatching("Defaults")?.LastChildMatching("Start", includeRemovals: false);
+				}
+
+				if (startNode != null)
+				{
+					var start = FieldLoader.GetValue<int>(startNode.Key, startNode.Value.Value);
+					if (start > 0)
+					{
+						length *= -1;
+						var add = Math.Max(start - length, 0) - 1;
+						var frames = new int[length > start ? start : length];
+						for (var i = frames.Length; i > 0; i--)
+							frames[^i] = i + add;
```

Comment:
Why so complicated?
```suggestion
						var add = Math.Max(start - length, 0);
						var frames = new int[length > start ? start : length];
						for (var i = 0; i < frames.Length; i++)
							frames[i] = i + add;
```

Reply:
because it needs to be in reverse. Your suggestion doesn't work

Reply:
```suggestion
						var frames = new int[length > start ? start : length];
						var add = Math.Max(start - length, 0) + frames.Length - 1;
						for (var i = 0; i < frames.Length; i++)
							frames[i] = -i + add;
```

Something like this would work, but tbf I find that more confusing

Reply:
Seems like you don't need to compute `add` that complicated at all then but can just roughly `(length > start ? start : length) - i` which will then also be more readable than "add".

Reply:
Made it simpler. Tbh I'm really not sure why bikeshedding is needed on update rules

Reply:
Just because they are update rules does not mean we should ignore writing good code.